actix-router/src/resource.rs

```diff
@@ -112,82 +137,264 @@ impl ResourceDef {
+            type Item = &'a str;
+
+            fn next(&mut self) -> Option<Self::Item> {
+                match &self.patterns {
+                    Patterns::Single(pattern) => {
+                        if self.done {
+                            return None;
+                        }
+
+                        self.done = true;
+                        Some(pattern.as_str())
+                    }
+                    Patterns::List(patterns) if patterns.is_empty() => None,
+                    Patterns::List(patterns) => match patterns.get(self.list_idx) {
+                        Some(pattern) => {
+                            self.list_idx += 1;
+                            Some(pattern.as_str())
+                        }
+                        None => {
+                            // fast path future call
+                            self.done = true;
+                            None
+                        }
+                    },
+                }
+            }
+
+            fn size_hint(&self) -> (usize, Option<usize>) {
+                match &self.patterns {
+                    Patterns::Single(_) => (1, Some(1)),
+                    Patterns::List(patterns) => (patterns.len(), Some(patterns.len())),
+                }
+            }
+        }
+
+        PatternIter {
+            patterns: &self.patterns,
+            list_idx: 0,
+            done: false,
+        }
     }
 
-    /// Check if path matches this pattern.
+    /// Returns `true` if `path` matches this resource.
+    ///
+    /// The behavior of this method depends on how the `ResourceDef` was constructed. For example,
+    /// static resources will not be able to match as many paths as dynamic and prefix resources.
+    ///
+    /// # Examples
+    /// ```
+    /// use actix_router::ResourceDef;
+    ///
+    /// // static resource
+    /// let resource = ResourceDef::new("/user");
+    /// assert!(resource.is_match("/user"));
+    /// assert!(!resource.is_match("/user/123"));
+    /// assert!(!resource.is_match("/foo"));
+    ///
+    /// // dynamic resource
+    /// let resource = ResourceDef::new("/user/{user_id}");
+    /// assert!(resource.is_match("/user/123"));
+    /// assert!(!resource.is_match("/user/123/stars"));
+    /// assert!(!resource.is_match("/foo"));
+    ///
+    /// // prefix resource
+    /// let resource = ResourceDef::prefix("/root");
+    /// assert!(resource.is_match("/root"));
+    /// assert!(resource.is_match("/root/leaf"));
+    /// assert!(!resource.is_match("/foo"));
+    ///
+    /// // TODO: dyn set resource
+    /// // TODO: tail segment resource
+    /// ```
     #[inline]
     pub fn is_match(&self, path: &str) -> bool {
         profile_method!(is_match);
 
+        // in effect this function could be expressed as:
+        // self.find_match(path).is_some()
+
         match self.pat_type {
             PatternType::Static(ref s) => s == path,
             PatternType::Prefix(ref s) => path.starts_with(s),
```

Comment:
```suggestion
            PatternType::Prefix(_) => self.find_match(s).is_some(),
```
This should have the logic to detect end of segment same as in `find_match`
Currently `ResourceDef::prefix("/user").is_match("/user123/abc")`!

Reply:
should have read them all, this is the discrepency I found also

in your mind, should we allow prefixes to do this or should they be used only for `/`-delimited "scopes"?

Reply:
I think this shouldn't be allowed. They should match full segments.

Reply:
okay, i was almost leaning the same way, i can't really think of a use case

Reply:
I've just found that full segment match is not required for dynamic prefixes in `find_match` and `match_path_checked` either. So in web the default service for `web::scope("/{domain}/user")` would be called on url `/xxx/user123`. This seems wrong to me.

Reply:
yeah i agree that's weird, good to start defining the router properly